diff --git a/utils/bazel/torch-mlir-overlay/BUILD.bazel b/utils/bazel/torch-mlir-overlay/BUILD.bazel
index 42abb127..7680af13 100644
--- a/utils/bazel/torch-mlir-overlay/BUILD.bazel
+++ b/utils/bazel/torch-mlir-overlay/BUILD.bazel
@@ -3,7 +3,7 @@
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 load("@llvm-project//mlir:tblgen.bzl", "gentbl_cc_library", "td_library")
-load("@com_github_bazelbuild_buildtools//buildifier:def.bzl", "buildifier")
+# load("@com_github_bazelbuild_buildtools//buildifier:def.bzl", "buildifier")
 
 package(
     default_visibility = [
@@ -13,9 +13,9 @@ package(
 
 # Format bazel BUILD and .bzl files with a standard convention
 # Invoke as `bazel run @torch-mlir//:buildifier`
-buildifier(
-    name = "buildifier",
-)
+# buildifier(
+#     name = "buildifier",
+# )
 
 # Torch Dialect/IR
 td_library(
